File: lib/dart/lib/src/protocol/t_protocol_decorator.dart
/// Licensed to the Apache Software Foundation (ASF) under one
/// or more contributor license agreements. See the NOTICE file
/// distributed with this work for additional information
/// regarding copyright ownership. The ASF licenses this file
/// to you under the Apache License, Version 2.0 (the
/// "License"); you may not use this file except in compliance
/// with the License. You may obtain a copy of the License at
///
/// http://www.apache.org/licenses/LICENSE-2.0
///
/// Unless required by applicable law or agreed to in writing,
/// software distributed under the License is distributed on an
/// "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
/// KIND, either express or implied. See the License for the
/// specific language governing permissions and limitations
/// under the License.

part of thrift;

/// Forward all operations to the wrapped protocol.  Used as a base class.
///
/// Adapted from the C# version.
class TProtocolDecorator extends TProtocol {
  final TProtocol _protocol;

  TProtocolDecorator(TProtocol protocol)
      : _protocol = protocol,
        super(protocol.transport);

  /// Write

  @override
  void writeMessageBegin(TMessage message) {
    _protocol.writeMessageBegin(message);
  }

  @override
  void writeMessageEnd() {
    _protocol.writeMessageEnd();
  }

  @override
  void writeStructBegin(TStruct struct) {
    _protocol.writeStructBegin(struct);
  }

  @override
  void writeStructEnd() {
    _protocol.writeStructEnd();
  }

  @override
  void writeFieldBegin(TField field) {
    _protocol.writeFieldBegin(field);
  }

  @override
  void writeFieldEnd() {
    _protocol.writeFieldEnd();
  }

  @override
  void writeFieldStop() {
    _protocol.writeFieldStop();
  }

  @override
  void writeMapBegin(TMap map) {
    _protocol.writeMapBegin(map);
  }

  @override
  void writeMapEnd() {
    _protocol.writeMapEnd();
  }

  @override
  void writeListBegin(TList list) {
    _protocol.writeListBegin(list);
  }

  @override
  void writeListEnd() {
    _protocol.writeListEnd();
  }

  @override
  void writeSetBegin(TSet set) {
    _protocol.writeSetBegin(set);
  }

  @override
  void writeSetEnd() {
    _protocol.writeSetEnd();
  }

  @override
  void writeBool(bool b) {
    _protocol.writeBool(b);
  }

  @override
  void writeByte(int b) {
    _protocol.writeByte(b);
  }

  @override
  void writeI16(int i16) {
    _protocol.writeI16(i16);
  }

  @override
  void writeI32(int i32) {
    _protocol.writeI32(i32);
  }

  @override
  void writeI64(int i64) {
    _protocol.writeI64(i64);
  }

  @override
  void writeDouble(double d) {
    _protocol.writeDouble(d);
  }

  @override
  void writeString(String str) {
    _protocol.writeString(str);
  }

  @override
  void writeBinary(Uint8List bytes) {
    _protocol.writeBinary(bytes);
  }

  /// Read
  @override
  TMessage readMessageBegin() => _protocol.readMessageBegin();
  @override
  void readMessageEnd() => _protocol.readMessageEnd();

  @override
  TStruct readStructBegin() => _protocol.readStructBegin();
  @override
  void readStructEnd() => _protocol.readStructEnd();

  @override
  TField readFieldBegin() => _protocol.readFieldBegin();
  @override
  void readFieldEnd() => _protocol.readFieldEnd();

  @override
  TMap readMapBegin() => _protocol.readMapBegin();
  @override
  void readMapEnd() => _protocol.readMapEnd();

  @override
  TList readListBegin() => _protocol.readListBegin();
  @override
  void readListEnd() => _protocol.readListEnd();

  @override
  TSet readSetBegin() => _protocol.readSetBegin();
  @override
  void readSetEnd() => _protocol.readSetEnd();

  @override
  bool readBool() => _protocol.readBool();

  @override
  int readByte() => _protocol.readByte();

  @override
  int readI16() => _protocol.readI16();

  @override
  int readI32() => _protocol.readI32();

  @override
  int readI64() => _protocol.readI64();

  @override
  double readDouble() => _protocol.readDouble();

  @override
  String readString() => _protocol.readString();

  @override
  Uint8List readBinary() => _protocol.readBinary();
}

File: lib/php/lib/Protocol/TProtocolDecorator.php
<?php

/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements. See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership. The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License. You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied. See the License for the
 * specific language governing permissions and limitations
 * under the License.
 *
 * @package thrift.protocol
 */

declare(strict_types=1);

namespace Thrift\Protocol;

use Thrift\Exception\TException;

/**
 * <code>TProtocolDecorator</code> forwards all requests to an enclosed
 * <code>TProtocol</code> instance, providing a way to author concise
 * concrete decorator subclasses. While it has no abstract methods, it
 * is marked abstract as a reminder that by itself, it does not modify
 * the behaviour of the enclosed <code>TProtocol</code>.
 *
 * @package Thrift\Protocol
 */
abstract class TProtocolDecorator extends TProtocol
{
    /**
     * Instance of protocol, to which all operations will be forwarded.
     */
    private TProtocol $concreteProtocol;

    /**
     * Constructor of <code>TProtocolDecorator</code> class.
     * Encloses the specified protocol.
     *
     * @param TProtocol $protocol All operations will be forward to this instance. Must be non-null.
     */
    protected function __construct(TProtocol $protocol)
    {
        parent::__construct($protocol->getTransport());
        $this->concreteProtocol = $protocol;
    }

    public function writeMessageBegin(string $name, int $type, int $seqid): int
    {
        return $this->concreteProtocol->writeMessageBegin($name, $type, $seqid);
    }

    public function writeMessageEnd(): int
    {
        return $this->concreteProtocol->writeMessageEnd();
    }

    /**
     * @throws TException on write error
     */
    public function writeStructBegin(string $name): int
    {
        return $this->concreteProtocol->writeStructBegin($name);
    }

    /**
     * @throws TException on write error
     */
    public function writeStructEnd(): int
    {
        return $this->concreteProtocol->writeStructEnd();
    }

    public function writeFieldBegin(string $fieldName, int $fieldType, int $fieldId): int
    {
        return $this->concreteProtocol->writeFieldBegin($fieldName, $fieldType, $fieldId);
    }

    public function writeFieldEnd(): int
    {
        return $this->concreteProtocol->writeFieldEnd();
    }

    public function writeFieldStop(): int
    {
        return $this->concreteProtocol->writeFieldStop();
    }

    public function writeMapBegin(int $keyType, int $valType, int $size): int
    {
        return $this->concreteProtocol->writeMapBegin($keyType, $valType, $size);
    }

    public function writeMapEnd(): int
    {
        return $this->concreteProtocol->writeMapEnd();
    }

    public function writeListBegin(int $elemType, int $size): int
    {
        return $this->concreteProtocol->writeListBegin($elemType, $size);
    }

    public function writeListEnd(): int
    {
        return $this->concreteProtocol->writeListEnd();
    }

    public function writeSetBegin(int $elemType, int $size): int
    {
        return $this->concreteProtocol->writeSetBegin($elemType, $size);
    }

    public function writeSetEnd(): int
    {
        return $this->concreteProtocol->writeSetEnd();
    }

    public function writeBool(bool $bool): int
    {
        return $this->concreteProtocol->writeBool($bool);
    }

    public function writeByte(int $byte): int
    {
        return $this->concreteProtocol->writeByte($byte);
    }

    public function writeI16(int $i16): int
    {
        return $this->concreteProtocol->writeI16($i16);
    }

    public function writeI32(int $i32): int
    {
        return $this->concreteProtocol->writeI32($i32);
    }

    public function writeI64(int $i64): int
    {
        return $this->concreteProtocol->writeI64($i64);
    }

    public function writeDouble(float $dub): int
    {
        return $this->concreteProtocol->writeDouble($dub);
    }

    public function writeString(string $str): int
    {
        return $this->concreteProtocol->writeString($str);
    }

    public function writeUuid(string $uuid): int
    {
        return $this->concreteProtocol->writeUuid($uuid);
    }

    public function readMessageBegin(?string &$name, ?int &$type, ?int &$seqid): int
    {
        return $this->concreteProtocol->readMessageBegin($name, $type, $seqid);
    }

    public function readMessageEnd(): int
    {
        return $this->concreteProtocol->readMessageEnd();
    }

    public function readStructBegin(?string &$name): int
    {
        return $this->concreteProtocol->readStructBegin($name);
    }

    public function readStructEnd(): int
    {
        return $this->concreteProtocol->readStructEnd();
    }

    public function readFieldBegin(?string &$name, ?int &$fieldType, ?int &$fieldId): int
    {
        return $this->concreteProtocol->readFieldBegin($name, $fieldType, $fieldId);
    }

    public function readFieldEnd(): int
    {
        return $this->concreteProtocol->readFieldEnd();
    }

    public function readMapBegin(?int &$keyType, ?int &$valType, ?int &$size): int
    {
        return $this->concreteProtocol->readMapBegin($keyType, $valType, $size);
    }

    public function readMapEnd(): int
    {
        return $this->concreteProtocol->readMapEnd();
    }

    public function readListBegin(?int &$elemType, ?int &$size): int
    {
        return $this->concreteProtocol->readListBegin($elemType, $size);
    }

    public function readListEnd(): int
    {
        return $this->concreteProtocol->readListEnd();
    }

    public function readSetBegin(?int &$elemType, ?int &$size): int
    {
        return $this->concreteProtocol->readSetBegin($elemType, $size);
    }

    public function readSetEnd(): int
    {
        return $this->concreteProtocol->readSetEnd();
    }

    public function readBool(?bool &$bool): int
    {
        return $this->concreteProtocol->readBool($bool);
    }

    public function readByte(?int &$byte): int
    {
        return $this->concreteProtocol->readByte($byte);
    }

    public function readI16(?int &$i16): int
    {
        return $this->concreteProtocol->readI16($i16);
    }

    public function readI32(?int &$i32): int
    {
        return $this->concreteProtocol->readI32($i32);
    }

    public function readI64(?int &$i64): int
    {
        return $this->concreteProtocol->readI64($i64);
    }

    public function readDouble(?float &$dub): int
    {
        return $this->concreteProtocol->readDouble($dub);
    }

    public function readString(?string &$str): int
    {
        return $this->concreteProtocol->readString($str);
    }

    public function readUuid(?string &$uuid): int
    {
        return $this->concreteProtocol->readUuid($uuid);
    }
}
